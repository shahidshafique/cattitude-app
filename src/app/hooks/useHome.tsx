import { useState, useCallback } from "react";
import { useBoundStore } from "../../store";
import { useFocusEffect } from "expo-router";

export const useHome = () => {
  const [loading, setLoading] = useState(true);
  const [refreshing, setRefreshing] = useState(false);
  const [error, setError] = useState<string | null>(null);

  const { cats, setCats } = useBoundStore();

  const loadCats = useCallback(async () => {
    setLoading(true);
    await setCats();
    setLoading(false);
  }, []);

  useFocusEffect(
    useCallback(() => {
      loadCats();
    }, []),
  );

  const onRefresh = () => {
    setRefreshing(true);
    setError(null);
    loadCats();
  };

  return { loading, refreshing, error, loadCats, onRefresh, cats };
};
