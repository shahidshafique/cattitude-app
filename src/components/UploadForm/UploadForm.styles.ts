import { StyleSheet } from "react-native";

export const styles = StyleSheet.create({
  container: {
    padding: 20,
  },
  selectButton: {
    backgroundColor: "#4CAF50",
    padding: 15,
    borderRadius: 8,
    alignItems: "center",
    marginBottom: 20,
  },
  uploadButton: {
    backgroundColor: "#2196F3",
    padding: 15,
    borderRadius: 8,
    alignItems: "center",
    marginTop: 20,
  },
  disabledButton: {
    backgroundColor: "#B0BEC5",
  },
  buttonText: {
    color: "white",
    fontSize: 16,
    fontWeight: "bold",
  },
  imagePreviewContainer: {
    alignItems: "center",
    marginVertical: 20,
  },
  imagePreview: {
    width: 300,
    height: 300,
    borderRadius: 8,
  },
  errorContainer: {
    backgroundColor: "#FFEBEE",
    padding: 10,
    borderRadius: 8,
    marginTop: 10,
  },
  errorText: {
    color: "#D32F2F",
  },
});
