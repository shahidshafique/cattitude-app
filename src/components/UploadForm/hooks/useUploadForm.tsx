import React, { useState } from "react";
import { Platform } from "react-native";
import * as ImagePicker from "expo-image-picker";
import { uploadImage } from "../../../services";
import { useRouter } from "expo-router";
import * as FileSystem from "expo-file-system";

export const useUploadForm = () => {
  const [image, setImage] = useState<string | null | undefined>(null);
  const [isUploading, setIsUploading] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const router = useRouter();

  const pickImage = async () => {
    setError(null);

    const permissionResult = await ImagePicker.requestMediaLibraryPermissionsAsync();

    if (!permissionResult.granted) {
      setError("Permission to access camera is required!");
      return;
    }

    const result = await ImagePicker.launchImageLibraryAsync({ mediaTypes: ["images"] });

    if (result.canceled || !result.assets || result.assets.length === 0) return;
  

    if (Platform.OS !== "web") {
        const imageUri = result.assets[0].uri;

        const fileExtension = imageUri.split(".").pop()?.toLowerCase();
        let mimeType = "image/jpeg"; 
        if (fileExtension === "png") {
          mimeType = "image/png";
        } else if (fileExtension === "jpg" || fileExtension === "jpeg") {
          mimeType = "image/jpeg";
        } else {
          setError("Unsupported image format. Please choose a PNG or JPEG image.");
          return;
        }
      const base64 = await FileSystem.readAsStringAsync(result.assets[0].uri, { encoding: FileSystem.EncodingType.Base64 });
      setImage(`data:${mimeType};base64,${base64}`);
    } else {
      setImage(result.assets[0].uri);
    }
  };

  const handleUpload = async () => {
    setIsUploading(true);
    setError(null);
    try {
      await uploadImage(image!);
      router.dismiss();
    } catch (error: any) {
      setError("Failed to upload image. Please try again.");
    } finally {
      setIsUploading(false);
    }
  };

  return {
    image,
    handleUpload,
    pickImage,
    error,
    isUploading,
  };
};
