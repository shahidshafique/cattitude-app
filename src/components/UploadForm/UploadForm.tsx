import React from "react";
import { View, Text, TouchableOpacity, Image, ActivityIndicator } from "react-native";
import { styles } from "./UploadForm.styles";
import { useUploadForm } from "./hooks/useUploadForm";

export const UploadForm = () => {
  const { image, pickImage, error, isUploading, handleUpload } = useUploadForm();

  return (
    <View style={styles.container}>
      <TouchableOpacity style={styles.selectButton} onPress={pickImage}>
        <Text style={styles.buttonText}>Select Image</Text>
      </TouchableOpacity>

      {image && (
        <View style={styles.imagePreviewContainer}>
          <Image source={{ uri: image }} style={styles.imagePreview} />
        </View>
      )}

      {error && (
        <View style={styles.errorContainer}>
          <Text style={styles.errorText}>{error}</Text>
        </View>
      )}

      <TouchableOpacity style={[styles.uploadButton, (!image || isUploading) && styles.disabledButton]} onPress={handleUpload} disabled={!image || isUploading}>
        {isUploading ? <ActivityIndicator color="#fff" /> : <Text style={styles.buttonText}>Upload Cat Image</Text>}
      </TouchableOpacity>
    </View>
  );
};

