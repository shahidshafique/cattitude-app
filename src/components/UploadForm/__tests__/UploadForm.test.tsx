import React from 'react';
import { render } from '@testing-library/react-native';
import { UploadForm } from '../UploadForm';
import { useUploadForm } from '../hooks/useUploadForm';

// Mock the hook
jest.mock('../hooks/useUploadForm');

describe('UploadForm', () => {
  // Helper to create a mock implementation for different states
  const setupMockHook = (overrides = {}) => {
    const mockHook = {
      image: null,
      pickImage: jest.fn(),
      error: null,
      isUploading: false,
      handleUpload: jest.fn(),
      ...overrides
    };
    
    (useUploadForm as jest.Mock).mockReturnValue(mockHook);
    return mockHook;
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly in initial state', () => {
    setupMockHook();
    const { toJSON } = render(<UploadForm />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with selected image', () => {
    setupMockHook({
      image: 'data:image/jpeg;base64,mockImageData'
    });
    const { toJSON } = render(<UploadForm />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with error message', () => {
    setupMockHook({
      error: 'Something went wrong'
    });
    const { toJSON } = render(<UploadForm />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly while uploading', () => {
    setupMockHook({
      image: 'data:image/jpeg;base64,mockImageData',
      isUploading: true
    });
    const { toJSON } = render(<UploadForm />);
    expect(toJSON()).toMatchSnapshot();
  });
});