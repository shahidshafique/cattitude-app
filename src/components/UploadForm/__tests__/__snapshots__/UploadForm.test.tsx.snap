// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadForm renders correctly in initial state 1`] = `
<View
  style={
    {
      "padding": 20,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#4CAF50",
        "borderRadius": 8,
        "marginBottom": 20,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Select Image
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#B0BEC5",
        "borderRadius": 8,
        "marginTop": 20,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Upload Cat Image
    </Text>
  </View>
</View>
`;

exports[`UploadForm renders correctly while uploading 1`] = `
<View
  style={
    {
      "padding": 20,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#4CAF50",
        "borderRadius": 8,
        "marginBottom": 20,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Select Image
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "marginVertical": 20,
      }
    }
  >
    <Image
      source={
        {
          "uri": "data:image/jpeg;base64,mockImageData",
        }
      }
      style={
        {
          "borderRadius": 8,
          "height": 300,
          "width": 300,
        }
      }
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#B0BEC5",
        "borderRadius": 8,
        "marginTop": 20,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <ActivityIndicator
      color="#fff"
    />
  </View>
</View>
`;

exports[`UploadForm renders correctly with error message 1`] = `
<View
  style={
    {
      "padding": 20,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#4CAF50",
        "borderRadius": 8,
        "marginBottom": 20,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Select Image
    </Text>
  </View>
  <View
    style={
      {
        "backgroundColor": "#FFEBEE",
        "borderRadius": 8,
        "marginTop": 10,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        {
          "color": "#D32F2F",
        }
      }
    >
      Something went wrong
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#B0BEC5",
        "borderRadius": 8,
        "marginTop": 20,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Upload Cat Image
    </Text>
  </View>
</View>
`;

exports[`UploadForm renders correctly with selected image 1`] = `
<View
  style={
    {
      "padding": 20,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#4CAF50",
        "borderRadius": 8,
        "marginBottom": 20,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Select Image
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "marginVertical": 20,
      }
    }
  >
    <Image
      source={
        {
          "uri": "data:image/jpeg;base64,mockImageData",
        }
      }
      style={
        {
          "borderRadius": 8,
          "height": 300,
          "width": 300,
        }
      }
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2196F3",
        "borderRadius": 8,
        "marginTop": 20,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Upload Cat Image
    </Text>
  </View>
</View>
`;
